ros2_control_demo_hardware/src/rrbot_with_gpios.cpp

```diff
@@ -0,0 +1,221 @@
+      info_.joints[i].name, hardware_interface::HW_IF_POSITION, &hw_commands_[i]));
+  }
+
+  return command_interfaces;
+}
+
+CallbackReturn RRBotSystemPositionOnlyHardware::on_activate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Starting ...please wait...");
+
+  for (int i = 0; i < hw_start_sec_; i++)
+  {
+    rclcpp::sleep_for(std::chrono::seconds(1));
+    RCLCPP_INFO(
+      rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "%.1f seconds left...",
+      hw_start_sec_ - i);
+  }
+
+  // command and state should be equal when starting
+  for (uint i = 0; i < hw_states_.size(); i++)
+  {
+    hw_commands_[i] = hw_states_[i];
+  }
+
+  RCLCPP_INFO(
+    rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "System Successfully started!");
+
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn RRBotSystemPositionOnlyHardware::on_deactivate(
+  const rclcpp_lifecycle::State & /*previous_state*/)
+{
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Stopping ...please wait...");
+
+  for (int i = 0; i < hw_stop_sec_; i++)
+  {
+    rclcpp::sleep_for(std::chrono::seconds(1));
+    RCLCPP_INFO(
+      rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "%.1f seconds left...",
+      hw_stop_sec_ - i);
+  }
+
+  RCLCPP_INFO(
+    rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "System successfully stopped!");
+
+  return CallbackReturn::SUCCESS;
+}
+
+hardware_interface::return_type RRBotSystemPositionOnlyHardware::read()
+{
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Reading...");
+
+  for (uint i = 0; i < hw_states_.size(); i++)
+  {
+    // Simulate RRBot's movement
+    hw_states_[i] = hw_states_[i] + (hw_commands_[i] - hw_states_[i]) / hw_slowdown_;
+    RCLCPP_INFO(
+      rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Got state %.5f for joint %d!",
+      hw_states_[i], i);
+  }
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Joints successfully read!");
+
+  return hardware_interface::return_type::OK;
+}
+
+hardware_interface::return_type RRBotSystemPositionOnlyHardware::write()
+{
+  RCLCPP_INFO(rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Writing...");
+
+  for (uint i = 0; i < hw_commands_.size(); i++)
+  {
+    // Simulate sending commands to the hardware
+    RCLCPP_INFO(
+      rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Got command %.5f for joint %d!",
+      hw_commands_[i], i);
+  }
+  RCLCPP_INFO(
+    rclcpp::get_logger("RRBotSystemPositionOnlyHardware"), "Joints successfully written!");
+
+  return hardware_interface::return_type::OK;
+}
+
+}  // namespace ros2_control_demo_hardware
+
+#include "pluginlib/class_list_macros.hpp"
+
+PLUGINLIB_EXPORT_CLASS(
+  ros2_control_demo_hardware::RRBotSystemPositionOnlyHardware, hardware_interface::SystemInterface)
```

Comment:
Apart some of the debugging messages I do not see really the difference between rrbot_with_gpios.cpp and rrbot_system_position_only.cpp ?

Reply:
Don't you think that the file name should be rrbot_system_with_gpios ?
And also that the class name  should be changed to RRbotSystemWithGpios ?